[PATCH] TOMOYO 2.5 backport patch for Linux 3.2

Apply this patch after overwriting security/tomoyo/ directory by doing

 $ wget https://www.kernel.org/pub/linux/kernel/v3.0/linux-3.5.tar.bz2
 $ tar -jxf linux-3.5.tar.bz2 --strip 1 linux-3.5/security/tomoyo/

.

Signed-off-by: Tetsuo Handa <penguin-kernel@I-love.SAKURA.ne.jp>
---
 security/tomoyo/tomoyo.c |   21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

--- linux-3.2.18.orig/security/tomoyo/tomoyo.c
+++ linux-3.2.18/security/tomoyo/tomoyo.c
@@ -186,7 +186,7 @@ static int tomoyo_path_unlink(struct pat
  * Returns 0 on success, negative value otherwise.
  */
 static int tomoyo_path_mkdir(struct path *parent, struct dentry *dentry,
-			     umode_t mode)
+			     int mode)
 {
 	struct path path = { parent->mnt, dentry };
 	return tomoyo_path_number_perm(TOMOYO_TYPE_MKDIR, &path,
@@ -234,7 +234,7 @@ static int tomoyo_path_symlink(struct pa
  * Returns 0 on success, negative value otherwise.
  */
 static int tomoyo_path_mknod(struct path *parent, struct dentry *dentry,
-			     umode_t mode, unsigned int dev)
+			     int mode, unsigned int dev)
 {
 	struct path path = { parent->mnt, dentry };
 	int type = TOMOYO_TYPE_CREATE;
@@ -319,14 +319,14 @@ static int tomoyo_file_fcntl(struct file
 }
 
 /**
- * tomoyo_file_open - Target for security_file_open().
+ * tomoyo_dentry_open - Target for security_dentry_open().
  *
  * @f:    Pointer to "struct file".
  * @cred: Pointer to "struct cred".
  *
  * Returns 0 on success, negative value otherwise.
  */
-static int tomoyo_file_open(struct file *f, const struct cred *cred)
+static int tomoyo_dentry_open(struct file *f, const struct cred *cred)
 {
 	int flags = f->f_flags;
 	/* Don't check read permission here if called from do_execve(). */
@@ -353,14 +353,17 @@ static int tomoyo_file_ioctl(struct file
 /**
  * tomoyo_path_chmod - Target for security_path_chmod().
  *
- * @path: Pointer to "struct path".
- * @mode: DAC permission mode.
+ * @dentry: Pointer to "struct dentry".
+ * @mnt:    Pointer to "struct vfsmount".
+ * @mode:   DAC permission mode.
  *
  * Returns 0 on success, negative value otherwise.
  */
-static int tomoyo_path_chmod(struct path *path, umode_t mode)
+static int tomoyo_path_chmod(struct dentry *dentry, struct vfsmount *mnt,
+			     mode_t mode)
 {
-	return tomoyo_path_number_perm(TOMOYO_TYPE_CHMOD, path,
+	struct path path = { mnt, dentry };
+	return tomoyo_path_number_perm(TOMOYO_TYPE_CHMOD, &path,
 				       mode & S_IALLUGO);
 }
 
@@ -510,7 +513,7 @@ static struct security_operations tomoyo
 	.bprm_set_creds      = tomoyo_bprm_set_creds,
 	.bprm_check_security = tomoyo_bprm_check_security,
 	.file_fcntl          = tomoyo_file_fcntl,
-	.file_open           = tomoyo_file_open,
+	.dentry_open         = tomoyo_dentry_open,
 	.path_truncate       = tomoyo_path_truncate,
 	.path_unlink         = tomoyo_path_unlink,
 	.path_mkdir          = tomoyo_path_mkdir,
